### PR TITLE
feat: add C implementation for `math/base/special/hyp2f1`

### LINKED ISSUE #122: RFC: evaluate the Gauss hypergeometric function
## Checklist

> Please ensure the following tasks are completed before filing an issue.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] If this is a general question, searched the [FAQ][faq] for an existing answer.
-   [x] If this is a feature request, the issue name begins with `RFC:`.

## Description

> Description of the issue (or feature request).

Add support for evaluating the [Gauss hypergeometric function](https://en.wikipedia.org/wiki/Hypergeometric_function).

Package: `@stdlib/math/base/special/hyp2f1`
Alias: `base.hypergeometric2f1` or `base.hyp2f1`

__WARNING__: this implementation is quite involved, but is useful as is used in other implementations, such as evaluating Legendre polynomials.

## Related Issues

> Does this issue (or feature request) have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this issue (or feature request)? This may include screenshots, references, stack traces, sample output, and/or implementation notes.

Reference implementations.

-   [Cephes](https://github.com/scipy/scipy/blob/2526df72e5d4ca8bad6e2f4b3cbdfbc33e805865/scipy/special/cephes/hyp2f1.c#L1)
-   [Mathematica](http://functions.wolfram.com/HypergeometricFunctions/Hypergeometric2F1/)

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[faq]: https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md

### PR DESCRIPTION
**Pull request by anandkaranubc**
Resolves https://github.com/stdlib-js/stdlib/issues/122

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the complete C implementation of the Gaussian hypergeometric function, along with updated benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/stdlib/issues/122

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+#include "stdlib/constants/float64/pinf.h"
+#include "stdlib/constants/float64/nan.h"
+#include <stdint.h>
+
+static const double MACHEP = 1.11022302462515654042e-16;
+static const double EPS = 1.0e-13;
+static const double ETHRESH = 1.0e-12;
+static const double MAX_ITERATIONS = 10000.0;
+
+static double hys2f1( double a, double b, const double c, const double x, double *loss );
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c:53]: This line was needed because of the circular dependency of `hys2f1` and `hyp2f1ra`

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js [hunk 2]
```diff
@@ -118,13 +118,13 @@ function hyp2f1( a, b, c, x ) {
 	var i;
 
 	err = 0.0;
+	s = 1.0 - x;
+	d = c - a - b;
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:122]: A minor bug fix where `d` was used before being assigned a value. This does not affect the test tolerances—only a minor change of a factor from `20` to `17` in one test.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js [hunk 3]
```diff
@@ -256,7 +256,7 @@ tape( 'the function correctly evaluates the hypergeometric function', function t
 			continue;
 		}
 		delta = abs( v - expected[ i ] );
-		tol = 20.0 * EPS * abs( expected[ i ] );
+		tol = 17.0 * EPS * abs( expected[ i ] );
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js:259]: Right here.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c (hunk 1 above)
Comment (anandkaranubc) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c:53]: For `hys2f1`, the first two parameters cannot be `const` because their values may be modified in the function depending on whether they satisfy a specific case.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+		b = a;
+		a = f;
+	}
+
+	if ( isNonPositiveInteger( b ) && ( stdlib_base_abs( b ) < stdlib_base_abs( a ) ) ) {
+		f = b;
+		b = a;
+		a = f;
+		intFlag = 1.0;
+	}
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c:197]: Follow up on the above [comment](https://github.com/stdlib-js/stdlib/pull/5348#discussion_r1964650615): This is the block where their values are being modified.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/test/fixtures/python/runner.py [hunk 5]
```diff
@@ -48,6 +48,7 @@
     (1.8, 7.4, -1.8, -1.01),
     (10, 1, -1.8, -0.99),
     (10, 7.4, -1.8, -0.99),
+    (0.5, 1 - 270.5, 1.5, 0.999**2),
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/test/fixtures/python/runner.py:51]: Because of compiler optimizations, this case was also added to the outliers list and removed from `# Edge Cases #1`.

It caused significantly higher divergence than other cases in the same test suite.

Comment (anandkaranubc) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/test/fixtures/python/runner.py:51]: I can confirm that if `CFLAGS="-ffp-contract=off" make install-node-addons NODE_ADDONS_PATTERN="@stdlib/math/base/special/hyp2f1"` is used, there is no need to do this.

**Comment by anandkaranubc** [comment]
Note: The exact same tolerance thresholds used for the JavaScript implementation can also be applied here if compiler optimizations are turned off.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js (hunk 2 above)
Comment (anandkaranubc) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:122]: This has already been taken in account for the C implementation.

**Comment by kgryte** [comment]
Thanks, @anandkaranubc, for the comments; those certainly help for review!

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*
+* ## Notice
+*
+* The original C code and copyright notice are from the [Cephes Mathematical Library]{@link https://www.netlib.org/cephes/}. The implementation has been modified for JavaScript.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]:
```suggestion
* The original C code and copyright notice are from the [Cephes Mathematical Library]{@link https://www.netlib.org/cephes/}. The implementation has been modified according to stdlib conventions.
```

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+
+static double hys2f1( double a, double b, const double c, const double x, double *loss );
+
+/**
+* Tests if a finite double-precision floating-point number is a "close enough" integer.
+*
+* @param x - input value
+* @returns boolean indicating whether the value is a "close enough" integer
+*/
+static double isInteger( const double x ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: @anandkaranubc Why are you returning a double here? Wouldn't it make more sense to return `bool`? Would need to add the `stdbool.h` include above. Same for L75.

Comment (anandkaranubc) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: Ohh yes! I should change it to bool. Sorry about that.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+static const double ETHRESH = 1.0e-12;
+static const double MAX_ITERATIONS = 10000.0;
+
+static double hys2f1( double a, double b, const double c, const double x, double *loss );
+
+/**
+* Tests if a finite double-precision floating-point number is a "close enough" integer.
+*
+* @param x - input value
+* @returns boolean indicating whether the value is a "close enough" integer
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]:
```suggestion
* @param x     input value
* @return      boolean indicating whether the value is a "close enough" integer
```

C Doxygen comments are not a 1:1 match with JSDoc. All comments need to be updated accordingly.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+
+	if ( x < -0.5 && !( negIntA || negIntB ) ) {
+		if ( b > a ) {
+			// Transformation based on AMS55 #15.3.4:
+			y = stdlib_base_pow( s, -a ) * hys2f1( a, c-b, c, -x/s, &err );
+		} else {
+			// Transformation based on AMS55 #15.3.5:
+			y = stdlib_base_pow( s, -b ) * hys2f1( c-a, b, c, -x/s, &err );
+		}
+		goto done;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: @anandkaranubc Avoid the use of `goto`. It is fine to have duplicate logic.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+	}
+
+	// The transformation for c - a or c - b negative integer (AMS55 #15.3.3)...
+	if ( d <= -1.0 && !( !isIntD && s < 0.0 ) && !( negIntA || negIntB ) ) {
+		return stdlib_base_pow( s, d ) * stdlib_base_hyp2f1( c-a, c-b, c, x );
+	}
+
+	// Check whether the series diverges...
+	if ( d <= 0.0 && x == 1.0 && !( negIntA || negIntB ) ) {
+		goto hypdiv;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: Each of these `goto` statements should be be removed as this sort of misdirection disrupts the linear flow of logic. Yes, they are a language feature, but, in general, they make code harder to reason about.

**Review by kgryte — CHANGES_REQUESTED** [review]
Overall, this is looking good. Just a couple of minor nits concerning the use of `goto` and return value types.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c (hunk 8 above)
Comment (anandkaranubc) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: What is the spacing criteria for this?

**Comment by anandkaranubc** [comment]
> Overall, this is looking good. Just a couple of minor nits concerning the use of `goto` and return value types.

Making those changes right away. Thanks for the feedback.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c (hunk 8 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: @anandkaranubc The descriptions just need to be aligned. Enough separation to be readable and distinguishable from the parameter name.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+* ## References
+*
+* -   AMS55 #15.2.10
+*
+* @param a		input value
+* @param b		input value
+* @param c		input value
+* @param x		input value
+* @param loss	starting loss of significance
+* @returns 		the function value
```
Comment (anandkaranubc) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: I do see them aligned locally. Not sure if they are supposed to look like this.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: @anandkaranubc Do not use tabs. Use spaces within comments.

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+static const double ETHRESH = 1.0e-12;
+static const double MAX_ITERATIONS = 10000.0;
+
+static double hys2f1( double a, double b, const double c, const double x, double *loss );
+
+/**
+* Tests if a finite double-precision floating-point number is a "close enough" integer.
+*
+* @param x		input value
+* @returns		boolean indicating whether the value is a "close enough" integer
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]: This should be `return` not `returns`

File: lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+	if ( ax == 1.0 ) {
+		if ( x > 0.0 ) {
+			if ( negIntCaOrCb ) {
+				if ( d >= 0.0 ) {
+					// Transformation based on AMS55 #15.3.3:
+					y = stdlib_base_pow( s, d ) * hys2f1( c - a, c - b, c, x, &err );
+					return y;
+				} else {
+					return STDLIB_CONSTANT_FLOAT64_PINF;
+				}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/hyp2f1/src/main.c]:
```suggestion
				}
				return STDLIB_CONSTANT_FLOAT64_PINF;
```

Else block is not necessary.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks for adding this, @anandkaranubc!

**Comment by anandkaranubc** [comment]
> LGTM. Thanks for adding this, @anandkaranubc!

Yayyy! Another feather in stdlib’s cap! 🎉
